### DISCUSSION TITLE
Hide cells in unstructured grid

### DISCUSSION THREAD

**Discussion by Raymark97** [pyvista/pyvista#5689] [Q&A]
I'm trying to plot a large 3d grid (coming from a cellar automata). The problem is only with performances, i want to hide unnecessary cells but  haven't fully understand [ghosting](https://docs.pyvista.org/version/stable/examples/02-plot/ghost-cells.html) because my UnstructuredGrid doesn't seems to have "facies" field.

here's my code:

```py
    cell_example = np.array(
              [
                  [0, 0, 0],
                  [1, 0, 0],
                  [1, 1, 0],
                  [0, 1, 0],
                  [0, 0, 1],
                  [1, 0, 1],
                  [1, 1, 1],
                  [0, 1, 1],
              ]
          )

    cells = []

    for i in range(size):
        for j in range(size):
            for k in range(size):
                if update_grid[i, j, k] == 1:
                    cells.append(cell_example + [i, j, k])
    print(len(cells))
    count = np.count_nonzero(update_grid == 1)
    cells_hex = np.arange(count*8).reshape([count, 8])
    if len(cells) != 0:
        points = np.vstack(cells).astype(np.float64)
        grid = pv.UnstructuredGrid({CellType.HEXAHEDRON: cells_hex}, points)
        ghosts = np.argwhere(grid["facies"] < 1.0)
        grid.remove_cells(ghosts)

        grid.plot(show_edges=True, clim=[0, 4])
    else:
        print("No cells alive")
```

this code fails because grid["facies"] does not exist
`KeyError: 'Data array (facies) not present in this dataset.'`

**Comment by MatthewFlamm** [comment]
I cannot run your example, can you give a minimally reproducing example (MRE)?

The [hide cell example](https://docs.pyvista.org/version/stable/examples/02-plot/ghost-cells.html) is currently broken, but for a different reason related to changes to in place transformations of DataSets.

But you have another problem to start.  The example uses `facies` cell data to find the cells to hide, but you don't have this, nor is it required to use.  If you already know the indices of the cells, you can do this directly.  See below as a simple example.

```py
import pyvista as pv

mesh = pv.ImageData(dimensions=(11, 11, 11)).cast_to_unstructured_grid()
# blank cells 0, 1, 2, 10, 11, 12
mesh.remove_cells([0, 1, 2, 10, 11, 12], inplace=True)  # note inplace=True difference from example
mesh.plot(show_edges=True, color='white')
```

![image](https://github.com/pyvista/pyvista/assets/39341281/55a26b02-00a5-40ea-8b01-0a7352ad26f9)


The `remove_cell` function only adds a cell_data filed `vtkGhostType` that is either `0` (not ghosted) or `1` (dtype `np.uint8`) (ghosted).  If you will be repeatedly modifying this, it will be faster to implement your own routine to manage this cell_data field rather than re-creating the mesh and using `remove_cell` over and over.  The function code is here.  However, it is more involved and requires interfacing with vtk directly.

https://github.com/pyvista/pyvista/blob/0acda5f428a5d734463e42be9cb9d4e9b0d422ad/pyvista/core/pointset.py#L101

**Comment by Raymark97** [comment]
here's my working code:

  ```
import pyvista as pv
  import numpy as np
  
  size = 128
  grid_size = (size, size, size)
  
  grid = np.random.choice([0, 1], size=grid_size)
  # here i evolve the grid
  mesh = pv.ImageData(dimensions=(size + 1, size + 1, size + 1)).cast_to_unstructured_grid()
  to_remove = [i for i, data in enumerate(grid.flatten()) if data == 0]
  
  #print(to_remove)
  mesh.remove_cells(to_remove, inplace=True)
  mesh.plot(show_edges=True, color='red')
```

but this is not what i wanted to achieve. Is there any way to hide cells that are not in sight to improve visualization performance? if i raise the size to 512 or 1024 the mesh becomes way too big
